### ISSUE TITLE
can we add new class to existing model?

### ISSUE THREAD

**Issue by meizano** [ultralytics/yolov5#1071]
## ❔Question
Can we add new class to existing model that have been trained.

## Additional context
Trained model consist label for : person, car
Adding class so the next training can contain: person, car, bike, class
without having to train the previous dataset

**Comment by glenn-jocher** [comment]
@meizano this question has been asked and answered before here, please review the issues.

**Comment by erolgerceker** [comment]
> @meizano this question has been asked and answered before here, please review the issues.

could you please add the link. i couldnt find it

**Comment by erolgerceker** [comment]
could you please add the link. i couldnt find it

**Comment by sivasai9398** [comment]
@erolgerceker did you find the link.?? I am also looking to add new classes to existing classes. @glenn-jocher Please help with this.

**Comment by glenn-jocher** [comment]
@sivasai9398 this is pretty simple. A model must be trained on all class you expect it to be able to detect during deployment. If you train it on classes a and b, and now you have class c, you should train it on classes a, b c.

**Comment by sivasai9398** [comment]
I want to use about 10 classes from coco dataset of pre trained yolov5 and 20 different classes which are not in coco dataset .for sure i will do custom model training using yolov5 for classes not in coco dataset. So is there any way to use both coco dataset classes and custom classes as a **single model** for my college project??.so what should be my approach?? @glenn-jocher

**Comment by glenn-jocher** [comment]
@sivasai9398 you can label your new classes starting from 80 and then simply append your new data to coco in your data.yaml. See GlobalWheat2020.yaml for an example of training on multiple datasets.

**Comment by sivasai9398** [comment]
So I should label my custom dataset from class 81 and train with any version of yolov5 model then of course I can detect classes from 81 can I able to detect classes from 0 to 80 also which are coco dataset classes?? @glenn-jocher Just to confirm because I had already have a dataset of custom 1500 images labelled from class 0 so I should change those all to detect all classes 0 to 80 and remaining classes.

**Comment by glenn-jocher** [comment]
@sivasai9398 you can train on both COCO and your data togethor:

```yaml
train:
 - coco/train/images
 - custom/train/images
```

**Comment by ckay016** [comment]
> @sivasai9398 you can train on both COCO and your data togethor:
> 
> ```yaml
> train:
>  - coco/train/images
>  - custom/train/images
> ```

@glenn-jocher  Will it work if I want to train only on person class from coco and in the custom dataset I have 2 other classes not present in yolo? 
So my classes will be ['person', 'class2', 'class3'] and train will be :
train:
  - coco/train2017.txt
  - custom/train/images

**Comment by glenn-jocher** [comment]
@ckay016 your classes would be [coco_0 ... coco_79, your_class_80],

**Comment by kashishnaqvi101** [comment]
> I want to use about 10 classes from coco dataset of pre trained yolov5 and 20 different classes which are not in coco dataset .for sure i will do custom model training using yolov5 for classes not in coco dataset. So is there any way to use both coco dataset classes and custom classes as a **single model** for my college project??.so what should be my approach?? @glenn-jocher

Hey did you fid how to do it?

**Comment by glenn-jocher** [comment]
@kashishnaqvi101 see https://community.ultralytics.com/t/how-to-combine-weights-to-detect-from-multiple-datasets

**Comment by kashishnaqvi101** [comment]
> @kashishnaqvi101 see https://community.ultralytics.com/t/how-to-combine-weights-to-detect-from-multiple-datasets

So what should be my yaml file? I tried doing this but its not working

```
download: https://ultralytics.com/assets/coco128.zip

train: 
      -datasets/coco128/images/train2017/  
      -../kashish/images/train/
val: 
     -datasets/coco128/images/train2017/
     -../kashish/images/val/
  


# Classes
nc: 81  # number of classes
names: ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
        'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
        'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
        'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
        'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
        'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
        'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
        'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
        'hair drier', 'toothbrush', 'money']  # class names
```
but it gives error, what should I put as paths of coco dataset and my custom dataset for training with multiple dataset

(my dataset is outside yolov5 folder on Colab)

**Comment by glenn-jocher** [comment]
@kashishnaqvi101 yaml requires a space after any dashes when creating lists. 

To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
         'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
         'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
         'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
         'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
         'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
         'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
         'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
         'hair drier', 'toothbrush' ]  # class names
```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>


Good luck 🍀 and let us know if you have any other questions!

**Comment by kashishnaqvi101** [comment]
> @kashishnaqvi101 yaml requires a space after any dashes when creating lists.
> 
> To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:
> 
> ### 1.1 Create dataset.yaml
> [COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
> 
> ```yaml
> # Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
> path: ../datasets/coco128  # dataset root dir
> train: images/train2017  # train images (relative to 'path') 128 images
> val: images/train2017  # val images (relative to 'path') 128 images
> test:  # test images (optional)
> 
> # Classes
> nc: 80  # number of classes
> names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
>          'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
>          'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
>          'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
>          'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
>          'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
>          'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
>          'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
>          'hair drier', 'toothbrush' ]  # class names
> ```
> 
> ### 1.2 Create Labels
> After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:
> 
> * One row per object
> * Each row is `class x_center y_center width height` format.
> * Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
> * Class numbers are zero-indexed (start from 0).
> 
> <img alt="Image Labels" width="800" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">
> 
> The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):
> 
> <img alt="" width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png">
> 
> ### 1.3 Organize Directories
> Organize your train and val images and labels according to the example below. YOLOv5 assumes `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
> 
> ```shell
> ../datasets/coco128/images/im0.jpg  # image
> ../datasets/coco128/labels/im0.txt  # label
> ```
> 
> <img alt="" width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png">
> 
> Good luck 🍀 and let us know if you have any other questions!

Can you just tell me, can I put two different directory paths one for coco and one for my custom dataset? 
would this work for multiple datasets training ?
**I have not added path variable but have simply given it entirely for coco.
For my custom dataset, just have given its paths till the images**
```
train: 
      -datasets/coco128/images/train2017  
      -kashish/images/train/
val: 
     -datasets/coco128/images/train2017
     -kashish/images/val/
```

**Comment by lzt9746** [comment]
I have several questions about add new classes to existing model:
1. Should I relabel my dataset from the beginning, or do I only need to append the new dataset which only includes new class labels to dataset.yaml like below:



```
train:
  - images/dataset_1  # old dataset, images could contain new classes but didn't labeled.
  - images/dataset_2   # new dataset, only include new classes labels
```
2. Should I create a new model when training starts using parameter --weights="",  or I trainning based on the existing model like parameter --weight="existing_model.pt"

**Comment by glenn-jocher** [comment]
@lzt9746 see https://community.ultralytics.com/t/how-to-combine-weights-to-detect-from-multiple-datasets

**Comment by stormsson** [comment]
I understand this is an old and closed issue, but looking for a solution for an issue like @meizano , search engine brought me directly here.

For future reference, this is another possible solution, but a bit more complex.

https://y-t-g.github.io/tutorials/yolov8n-add-classes/
